### PR TITLE
Add pugblicgen

### LINKED ISSUE #15: New models will include the entire mgo stack when included
To prevent this, we could write a tool that will generate public models from the real ones

### PR DESCRIPTION
**Pull request by johnsudaar**
Fix #15

### CODE REVIEW

**Comment by johnsudaar** [comment]
Sample use of this package:
```
package main

import (
	"fmt"
	"os"

	"github.com/Scalingo/go-utils/publicgen"
	"github.com/Scalingo/scalingo-acme-agent/models"
)

func main() {
	err := publicgen.GeneratePublicModels(publicgen.GeneratorParams{
		OutputFile:    "publicapi/publicapi.go",
		OutputPackage: "publicapi",
		Types: []interface{}{
			models.Domain{},
			models.Certificate{},
		},
	})
	if err != nil {
		panic(err)
	}
}
```

**Comment by johnsudaar** [comment]
2018-03-06_1387x1133_scrot

File: publicgen/reflect.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+			continue
+		}
+
+		name := curField.Name
+		if unicode.IsLower([]rune(name)[0]) {
+			continue
+		}
+
+		json := curField.Tag.Get("json")
+		slice := false
```
Comment (EtienneM) [inline publicgen/reflect.go]: slice -> `isSlice`?

File: publicgen/reflect.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		typ := curField.Type
+
+		if curField.Type.Kind() == reflect.Slice {
+			slice = true
+			typ = curField.Type.Elem()
+		}
+
+		pointer := false
```
Comment (EtienneM) [inline publicgen/reflect.go]: pointer -> isPointer?

File: publicgen/reflect.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+		name := curField.Name
+		if unicode.IsLower([]rune(name)[0]) {
+			continue
+		}
+
+		json := curField.Tag.Get("json")
+		slice := false
+		typ := curField.Type
+
+		if curField.Type.Kind() == reflect.Slice {
```
Comment (EtienneM) [inline publicgen/reflect.go]: curField.Type -> `typ`

File: publicgen/reflect.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+			slice = true
+			typ = curField.Type.Elem()
```
Comment (EtienneM) [inline publicgen/reflect.go]: curField.Type -> `typ`

File: publicgen/reflect.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+			typ = typ.Elem()
+			pointer = true
+		}
+
+		pkg := typ.PkgPath()
+		typePrefix := ""
+
+		if pkg != "" {
+			path := strings.Split(pkg, "/")
+			typePrefix = path[len(path)-1] + "."
```
Comment (EtienneM) [inline publicgen/reflect.go]: Would be nice to have a comment with an example to better understand this string manipulation

File: publicgen/reflect.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
+package publicgen
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"unicode"
+)
+
+var typeStruct = map[string]string{
```
Comment (EtienneM) [inline publicgen/reflect.go]: I don't really like the name `typeStruct`. It seems that you use this to map type to the aliased type. Maybe something like `aliasedTypeMap` is better?

File: publicgen/ast.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+	return &ast.GenDecl{
+		Rparen: 1,
+		Tok:    token.IMPORT,
+		Specs:  importDecls,
+		Lparen: 1,
+	}
+}
+
+func newAST(packageName string, structs map[string][]field) *ast.File {
```
Comment (EtienneM) [inline publicgen/ast.go]: Add a comment to explain what it does, does not look straight forward

**Comment by Soulou** [comment]
In the last example screenshot, some tags were missing for some `time.Time`, why?

**Comment by johnsudaar** [comment]
They were not any defined in the base model.

**Comment by Soulou** [comment]
So why are there present in the public API, there is something I don't get

**Comment by johnsudaar** [comment]
The original model was like:

```
type Certificate struct{
  ID              bson.ObjectId `json:"id"`
  CreatedAt time.Time        `json:"created_at"`
// ...
  CerticateQueriedAt time.Time
//....
}
```
Note that the json tag is missing on the `CertificateQueriedAt` so i the generator wont add one.

**Pull request merged**